allennlp/modules/token_embedders/pretrained_transformer_embedder.py

```diff
@@ -38,4 +43,8 @@ def forward(
 
         Shape: [batch_size, num_wordpieces, embedding_size].
         """
-        return self.transformer_model(input_ids=token_ids, attention_mask=mask)[0]
+        if self.transformer_model.embeddings.token_type_embeddings.num_embeddings == 1:
```

Comment:
Could you add a comment explaining this condition? At a glance it looks like the model does in fact have type ids given that it has `num_embeddings` > 0.

Reply:
It seems like a token indexer which uses a model which does not have `type_ids` should not be producing them, rather than catching it here in the token embedder. I.e this `if` statement should be unnecessary if the token indexer just didn't produce token_type_ids for relevant models.

Reply:
`type_ids` doesn't have to come from the indexer. In my model, it comes from the dataset reader. That model uses `type_ids` to distinguish the context (which may contain the answer) from question and special tokens (which may not).

I could buy the argument that then the reader should not produce these, but it makes the code much more convenient if I can handle it in one place. Most of the time, `type_ids` are just a "bonus" feature that the model may or may not take advantage of, and RoBERTa does not.

For some models though, `type_ids` carry critical information. In those cases, it would be an error to use RoBERTa at all. I've built such a model before. To get it to work, I trained the type embeddings from scratch during fine-tuning (as they are discussing here: https://github.com/huggingface/transformers/issues/1234). It works, and beats the other transformers on the task. It would be a neat feature if we could offer RoBERTa with type ids, but that's out of scope for this PR.

On balance, I'm leaning towards moving this into the reader, but let me know what you think.

Reply:
I'd defer to you and Mark on the reader question. But if you do want to leave this check here, please comment the condition. It doesn't make any sense to me.

Reply:
I changed this so that it throws an exception if your type ids are too big for the embeddings. It's a runtime check that will have a minuscule impact on performance, but when you get it wrong, the error message is really cryptic. e1be452cc12a38034ce19e3654a36c0f9b03058b

Reply:
I think the prior approach is actually better, as it uses fewer tensor operations.  There's no need for the additional operation here.  I'd agree with "minuscule" if it were one-time, but this is at every call to `forward`.

@brendan-ai2, the prior check worked because it's checking the number of embeddings in an embedding matrix (1, meaning that everything has the same, non-informative type id), not the number of embedding matrices.

Reply:
Ah, thanks @matt-gardner !

Reply:
I'm happy to take it out. Consider though that the error message you get when this is wrong is `RuntimeError: cublas runtime error : library not initialized at /tmp/pip-req-build-ocx5vxk7/aten/src/THC/THCGeneral.cpp:216`.

Reply:
I did a little benchmarking for this. For `roberta-base`, the `forward()` call will take 316 ms (batch size 16). The `max()` call will take 68.5 µs. In other words, it adds 0.02% to the runtime of the `forward()` call.

Reply:
Fair enough, it's up to you. I wasn't trying to stop you from doing something, just express the general ideal that minimizing tensor operations is good, both for runtime and for memory. If you're confident it doesn't matter in this case, great.